File: jjazz-pure-core/src/main/java/org/jjazz/purecore/phrase/NoteEvent.java
package org.jjazz.purecore.phrase;

import static com.google.common.base.Preconditions.checkArgument;
import org.jjazz.purecore.harmony.Note;
import org.jjazz.purecore.util.FloatRange;
import java.util.Objects;

/**
 * A Note with a position in beats. Extracted from JJazzLab NoteEvent.
 */
public class NoteEvent extends Note {

    private final float position;

    public NoteEvent(int pitch, float duration, int velocity, float posInBeats, Accidental acc) {
        super(pitch, duration, velocity, acc);
        if (posInBeats < 0) throw new IllegalArgumentException("posInBeats=" + posInBeats);
        this.position = posInBeats;
    }

    public NoteEvent(int pitch, float duration, int velocity, float posInBeats) {
        this(pitch, duration, velocity, posInBeats, Accidental.FLAT);
    }

    public NoteEvent(Note n, float posInBeats) {
        this(n.getPitch(), n.getDurationInBeats(), n.getVelocity(), posInBeats, n.getAccidental());
    }

    public float getPositionInBeats() { return position; }

    public FloatRange getBeatRange() {
        return new FloatRange(position, position + getDurationInBeats());
    }

    public boolean isBefore(NoteEvent other) { return position < other.position; }

    /** Create a new NoteEvent with some fields modified. Pass negative/null to keep original. */
    public NoteEvent setAll(int pitch, float duration, int velocity, float posInBeats, Accidental acc) {
        return new NoteEvent(
            pitch < 0 ? getPitch() : pitch,
            duration < 0 ? getDurationInBeats() : duration,
            velocity < 0 ? getVelocity() : velocity,
            posInBeats < 0 ? getPositionInBeats() : posInBeats,
            acc == null ? getAccidental() : acc
        );
    }

    public NoteEvent setPitch(int newPitch) { return setAll(newPitch, -1, -1, -1, null); }
    public NoteEvent setDuration(float newDur) { return setAll(-1, newDur, -1, -1, null); }
    public NoteEvent setVelocity(int newVel) { return setAll(-1, -1, newVel, -1, null); }
    public NoteEvent setPosition(float newPos) { return setAll(-1, -1, -1, newPos, null); }

    @Override
    public NoteEvent clone() {
        return new NoteEvent(getPitch(), getDurationInBeats(), getVelocity(), position, getAccidental());
    }

    @Override
    public int compareTo(Note n) {
        if (n == this) return 0;
        if (n instanceof NoteEvent ne) {
            int res = Float.compare(position, ne.position);
            if (res == 0) res = super.compareTo(n);
            return res;
        }
        return super.compareTo(n);
    }

    @Override
    public boolean equals(Object o) { return this == o; }

    @Override
    public int hashCode() { return System.identityHashCode(this); }

    @Override
    public String toString() {
        return String.format("[%s, p=%.3f, d=%.3f, v=%d]",
            toPianoOctaveString(), position, getDurationInBeats(), getVelocity());
    }

    public static String saveAsString(NoteEvent ne) {
        return Note.saveAsString(ne, true) + ":" + ne.position;
    }
}

File: jjazz-engine/src/phrase/note_event.rs
use crate::harmony::note::{Note, Accidental};
use serde::{Serialize, Deserialize};

/// A Note with a position in beats on a timeline.
#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
pub struct NoteEvent {
    pub pitch: u8,
    pub duration: f32,
    pub velocity: u8,
    pub position: f32,
    pub accidental: Accidental,
}

impl NoteEvent {
    pub fn new(pitch: u8, duration: f32, velocity: u8, position: f32) -> Self {
        Self { pitch, duration, velocity, position, accidental: Accidental::Flat }
    }

    pub fn from_note(note: &Note, position: f32) -> Self {
        Self { pitch: note.pitch, duration: note.duration, velocity: note.velocity, position, accidental: note.accidental }
    }

    pub fn end_position(&self) -> f32 { self.position + self.duration }

    pub fn set_pitch(&self, pitch: u8) -> Self { Self { pitch, ..*self } }
    pub fn set_duration(&self, duration: f32) -> Self { Self { duration, ..*self } }
    pub fn set_velocity(&self, velocity: u8) -> Self { Self { velocity, ..*self } }
    pub fn set_position(&self, position: f32) -> Self { Self { position, ..*self } }

    pub fn piano_octave_string(&self) -> String {
        Note::with_all(self.pitch, self.duration, self.velocity, self.accidental).piano_octave_string()
    }
}
